src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -199,33 +199,65 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
                 throw new ArgumentOutOfRangeException(nameof(serverAddresses));
             }
 
-            // Try each address in turn, and return the socket opened for the first one that works.
-            ExceptionDispatchInfo lastException = null;
+            // Connect to each IP address in parallel, and return the socket opened for the first one that works.
+            List<Task> connectTasks = new List<Task>(serverAddresses.Length);
+
+            object socketOpenSetResultLock = new object();
+            Socket resultSocket = null;
+            ExceptionDispatchInfo lastSocketConnectEx = null;
+
+            CancellationTokenSource cancelTokenSource = new CancellationTokenSource();
+            CancellationToken cancelToken = cancelTokenSource.Token;
+
             foreach (IPAddress address in serverAddresses)
             {
-                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                 try
                 {
-                    await socket.ConnectAsync(address, port).ConfigureAwait(false);
-                    return socket;
+                    Task socketTask = socket.ConnectAsync(address, port).ContinueWith(
+                        t =>
+                        {
+                            lock (socketOpenSetResultLock)
+                            {
+                                if (!cancelToken.IsCancellationRequested && socket.Connected)
+                                {
+                                    // Successful socket connection, so cancel other pending tasks
+                                    resultSocket = socket;
+                                    cancelTokenSource.Cancel();
+                                    return;
+                                }
+                            }
+
+                            // Didn't return from setting result, so dispose this socket
+                            socket.Dispose();
```

Comment:
When one connection completes, it'll request cancellation, and the rest of the continuations will get canceled, which means the socket for those others won't be disposed.  Is that by design?